File: desafio1/respuesta/respuesta.py
import json
from collections import Counter

# tuples [x, y] 
# x, y days of the year [0,364]
# x first day seen, y day not seen
# calculate the density of titans for each day 
# and get the first day with the lowest density of titans
# (use the least amount of loops)

# questions: memory restrictions?
# questions: time restrictions?
# questions: el output debe especificar el día mes y año? ej día (369, 4) estaría bien?

# read the file and save the data in a list
def read_file(file_name):
    data = []
    file = open(file_name, 'r')
    raw_data = json.load(file)
    for x, y in raw_data:
        data.extend(list(range(x, y)))
    file.close()
    return data

# calculate the density of titans for each day
def least_occurrences(data, condition=None):
    counter = Counter(data)
    if callable(condition):
        counter = Counter({k: v for k, v in counter.items() if condition(k)})
    least_common = counter.most_common()[-1]
    least_amount = least_common[1]
    # if there are more than one day with the same density, get the first one
    least_common_days = [k for k, v in counter.most_common() if v == least_amount]

    return min(least_common_days), least_amount

days_with_titans = read_file('desafio1\last_year.json')

# ignore the months of january and december
condition = lambda x: x%365 not in range(0, 31) and x%365 not in range(334, 365)

least_common = least_occurrences(days_with_titans, condition)
print(least_common[0])

File: desafio1/respuesta/respuesta.php
<?php

/**
 * Reads a file and returns an array with the days with titans
 * If the file is not valid, it returns an empty array
 * @param string $file_name Name of the file to read
 * @return array Array with the days with titans
 */
function read_file($file_name)
{
    $data = [];
    $file = fopen($file_name, 'r');
    $raw_data = json_decode(fread($file, filesize($file_name)), true);
    fclose($file);

    if (!is_array($raw_data)) {
        print_r('The data is not valid: ' . json_encode($raw_data) . "\n");
        return [];
    }
    foreach ($raw_data as $item) {
        if (!is_array($item) || count($item) != 2) {
            print_r('The data is not valid: ' . json_encode($item) . "\n");
            return [];
        }
        $x = $item[0];
        $y = $item[1];
        $data = array_merge($data, range($x, $y - 1));
    }

    return $data;
}

/**
 * Returns the nearest day with the least titans and the amount of times it appears
 * If the array is empty, it returns [INF, INF]
 * @param array $data Data with the days with titans
 * @param callable|null $condition Condition to filter the days
 * @return array Array with the nearest day with the least titans and the amount of times it appears
 */
function get_nearest_day($data, $condition = null)
{
    if (empty($data)) {
        return [INF, INF];
    }
    $counter = array_count_values($data);

    $nearest_day = INF;
    $least_amount = INF;

    foreach ($counter as $day => $amount) {
        if (is_callable($condition) && !$condition($day)) {
            continue;
        }
        if ($amount > $least_amount) {
            continue;
        }
        if ($amount < $least_amount) {
            $least_amount = $amount;
            $nearest_day = $day;
        } elseif ($amount == $least_amount && $day < $nearest_day) {
            $nearest_day = $day;
        }
    }

    return [$nearest_day, $least_amount];
}

$days_with_titans = read_file('../last_year.json');

// Ignore the months of January and December
$condition = function ($day) {
    $year_day = $day % 365;
    return !((0 <= $year_day && $year_day < 31) || (334 <= $year_day && $year_day < 365));
};

$nearest_day = get_nearest_day($days_with_titans, $condition);
print_r($nearest_day[0]);
